staging/src/k8s.io/apiserver/pkg/storage/cacher/store_btree.go

```diff
@@ -385,3 +393,79 @@ func (i *indexer) delete(key, value string, index map[string]map[string]*storeEl
 		delete(index, value)
 	}
 }
+
+// storeSnapshotter caches snapshots of store created by cloning the store.
+type storeSnapshotter struct {
+	mux       sync.RWMutex
+	snapshots *btree.BTreeG[rvSnapshot]
+}
+
+type rvSnapshot struct {
+	resourceVersion uint64
+	snapshot        orderedLister
+}
+
+func newStoreSnapshotter() *storeSnapshotter {
+	s := &storeSnapshotter{}
+	s.reset()
+	return s
+}
+
+func (s *storeSnapshotter) Reset() {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	s.reset()
+}
+
+func (s *storeSnapshotter) reset() {
+	s.snapshots = btree.NewG[rvSnapshot](btreeDegree, func(a, b rvSnapshot) bool {
+		return a.resourceVersion < b.resourceVersion
+	})
+}
+
+func (s *storeSnapshotter) GetLessOrEqual(rv uint64) (orderedLister, bool) {
+	s.mux.RLock()
+	defer s.mux.RUnlock()
+
+	var result *rvSnapshot
+	s.snapshots.DescendLessOrEqual(rvSnapshot{resourceVersion: rv}, func(rvs rvSnapshot) bool {
+		result = &rvs
+		return false
+	})
+	if result == nil {
+		return nil, false
+	}
+	return result.snapshot, true
+}
+
+func (s *storeSnapshotter) HasLessOrEqual(rv uint64) bool {
```

Comment:
> In the PoC I used it for shouldDelegateList, however it's just for convenience when calling and maybe in the future we could cache minimal snapshot rv allowing for constant time lookup. We could remove it to simplify testing.

Yeah - storing minimal rv sounds like a better approach - so let's remove it if we don't have a clear usecase of it.